### PR TITLE
22 upload translation file

### LINKED ISSUE #22: Upload translation file
The purpose of this endpoint is to let consumers upload translations in the original language to REST API from the shell using `curl`, `wget` etc.

Upload file with additional fields:
* labels (map[string]string)
* type (enum: [ng-localise, ngx-translate, go])

**Notes.** 
* Let's rename `go messages` with simply `go` everywhere.
* no storage yet

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**

### CODE REVIEW

File: .vscode/extensions.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "recommendations": [
```
Comment (VladislavsPerkanuks) [inline .vscode/extensions.json:2]: https://dev.to/askrishnapravin/recommend-vs-code-extensions-to-your-future-teammates-4gkb

**Comment by VladislavsPerkanuks** [comment]
As I understood, clients will have to encode the data to bytes themselves, e.g. when requesting using curl
```bash
curl --location --request POST 'localhost:8080/v1/upload/translationfile' \
--header 'Content-Type: application/json' \
--data-raw '{
"data" : "ewoiYSI6ImIiLAoiYS5jIiA6ICJkIiwKInZsYWRpc2xhdnMiOiJwZXJrYW5rcyIKfQ=="
}'
```

**Review by DarjaLeonova — APPROVED** [review]

File: api/translate/translate/v1/translate.proto [hunk 2]
```diff
@@ -16,12 +23,20 @@ message Messages {
...
-message DummyResponse {}
+message UploadTranslationFileResponse {}
 
 service TranslateService {
-  rpc Dummy(DummyRequest) returns (DummyResponse) {
-    option (google.api.http) = {get: "/v1/dummy"};
+  rpc UploadTranslationFile(UploadTranslationFileRequest) returns (UploadTranslationFileResponse) {
+    option (google.api.http) = {
+      put: "/v1/translations/{translationID}"
+      body: "*"
```
Comment (VladislavsPerkanuks) [inline api/translate/translate/v1/translate.proto:39]: Example curl
```bash
curl --location --request PUT 'localhost:8080/v1/translations/asd' \
--header 'Content-Type: application/json' \
--data-raw '{
    "labels": {
        "aaa": "bbb",
        "version": "!"
    },
    "type": "NGX_TRANSLATE",
    "data": "{\"vlad\":\"perk\",\"beep\":{\"hey\":\"world\"}}"
}'
```

File: pkg/convert/go.go [hunk 3]
```diff
@@ -19,7 +19,7 @@ func FromGoMessages(m model.Messages) ([]byte, error) {
 	return msg, nil
 }
 
-func ToGoMessages(b []byte) (model.Messages, error) {
+func FromGo(b []byte) (model.Messages, error) {
```
Comment (VladislavsPerkanuks) [inline pkg/convert/go.go:22]: Refactored so it's consistent with other converts

From: bytes→messages
To: messages→bytes

File: pkg/translate/translate.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+) (*pb.UploadTranslationFileResponse, error) {
+	var (
+		messages model.Messages
+		err      error
+	)
+
+	switch req.GetType() {
+	default:
+		return nil, status.Errorf(codes.Internal, "")
+	case pb.Type_TYPE_UNSPECIFIED:
```
Comment (VladislavsPerkanuks) [inline pkg/translate/translate.go]: Not good-looking, but it's required by buf-lint to pass. It requires for enum in.proto files to be prefixed with "TYPE_"

File: api/translate/translate/v1/translate.proto [hunk 5]
```diff
@@ -16,12 +23,20 @@ message Messages {
...
-message DummyResponse {}
+message UploadTranslationFileResponse {}
 
 service TranslateService {
-  rpc Dummy(DummyRequest) returns (DummyResponse) {
-    option (google.api.http) = {get: "/v1/dummy"};
+  rpc UploadTranslationFile(UploadTranslationFileRequest) returns (UploadTranslationFileResponse) {
+    option (google.api.http) = {
+      put: "/v1/translations/{translation_id}"
+      body: "data"
```
Comment (VladislavsPerkanuks) [inline api/translate/translate/v1/translate.proto:40]: Example curl:
```bash
curl --location -g --request PUT 'localhost:8080/v1/translations/asd?type=NGX_TRANSLATE&labels[version]=1&labels[date]=2023-01-04' \
--header 'Content-Type: application/json' \
--data-raw '"{\"vlad\":\"perk\",\"beep\":{\"hey\":\"world\"}}"
'
```

File: .vscode/extensions.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+    "emeraldwalk.runonsave"
+  ]
```
Comment (janishorsts) [inline .vscode/extensions.json]:
```suggestion
  "recommendations": [
    "bufbuild.vscode-buf",
    "earthly.earthfile-syntax-highlighting",
    "emeraldwalk.runonsave",
    "mikestead.dotenv"
  ]
```

File: pkg/translate/translate.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+		return nil, status.Errorf(codes.InvalidArgument, "Type is missing")
+	case pb.Type_NG_LOCALISE:
+		messages, err = convert.FromNgJson([]byte(req.Data))
+	case pb.Type_NGX_TRANSLATE:
+		messages, err = convert.FromNgxTranslate([]byte(req.Data))
+	case pb.Type_GO:
+		messages, err = convert.FromGo([]byte(req.Data))
+	}
```
Comment (janishorsts) [inline pkg/translate/translate.go]: Functions are first-class citizens in golang. Define `type From = func ([]byte) (model.Messages, error)` in the `convert` package.

```suggestion

	var from convert.From
	
	switch v := req.GetType(); v {
	default:
		return nil, status.Errorf(codes.Unimplemented, "Type '%s' is unimplemented", pb.Type_name[v])
	case pb.Type_TYPE_UNSPECIFIED:
		return nil, status.Errorf(codes.InvalidArgument, "Type is missing")
	case pb.Type_NG_LOCALISE:
		from = convert.FromNgJson
	case pb.Type_NGX_TRANSLATE:
		from = convert.FromNgxTranslate
	case pb.Type_GO:
		from = convert.FromGo
	}
	
	messages, err := from([]byte(req.Data)
	// ...
```

File: pkg/translate/translate.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+	// todo Proper error handling with status codes
+	if err != nil {
+		return nil, fmt.Errorf("convert to messages: %w", err)
```
Comment (janishorsts) [inline pkg/translate/translate.go]:
```suggestion
		return nil, status.Errorf(codes.InvalidArgument, "parse data: %w", err)
```

File: pkg/translate/translate.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	}
+
+	// Save to DB/FS...
+	_ = messages
```
Comment (janishorsts) [inline pkg/translate/translate.go:67]:
```suggestion
	messages.Labels = req.GetLabels()
	_ = messages
```

**Review by janishorsts — CHANGES_REQUESTED** [review]

File: pkg/translate/translate.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+	ctx context.Context,
+	req *pb.UploadTranslationFileRequest,
+) (*pb.UploadTranslationFileResponse, error) {
+	var (
+		reqTranslationID = req.GetTranslationId()
+		reqData          = req.GetData()
+		reqLabels        = req.GetLabels()
+		reqType          = req.GetType()
+		reqLanguage      = req.GetLanguage()
+	)
```
Comment (VladislavsPerkanuks) [inline pkg/translate/translate.go:31]: Moved all request variables to one place, I think it's tidier that way + getters are called only once per variable

File: pkg/translate/translate.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	switch reqType {
+	case pb.Type_TYPE_UNSPECIFIED:
+		return nil, status.Errorf(codes.InvalidArgument, "type is missing")
+	case pb.Type_NG_LOCALISE:
+		from = convert.FromNgJson
+	case pb.Type_NGX_TRANSLATE:
+		from = convert.FromNgxTranslate
+	case pb.Type_GO:
+		from = convert.FromGo
+	}
```
Comment (VladislavsPerkanuks) [inline pkg/translate/translate.go:57]: Deleted `default` as it is impossible to get in that case. Neither gRPC request allows calling with unsupported type, nor REST.

File: pkg/convert/go.go [hunk 12]
```diff
@@ -8,7 +8,9 @@ import (
 	"golang.org/x/text/message/pipeline"
 )
 
-func FromGoMessages(m model.Messages) ([]byte, error) {
+type From = func([]byte) (model.Messages, error)
```
Comment (janishorsts) [inline pkg/convert/go.go]: It applies to all `func convert.From<name>()`. Why in `go.go`? Why not `convert.go` to emphasize its context of use?

**Review by janishorsts — CHANGES_REQUESTED** [review]

File: pkg/convert/go.go (hunk 12 above)
Comment (VladislavsPerkanuks) [reply pkg/convert/go.go]: Over sighted

**Review by janishorsts — APPROVED** [review]

**Pull request merged**